hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchemaComparatorForSchemaEvolution.java

```diff
@@ -0,0 +1,375 @@
+
+    if (fields1.size() != fields2.size()) {
+      return false;
+    }
+
+    for (int i = 0; i < fields1.size(); i++) {
+      if (!fieldEquals(fields1.get(i), fields2.get(i))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected boolean validateField(HoodieSchemaField f1, HoodieSchemaField f2) {
+    if (!f1.name().equals(f2.name())) {
+      return false;
+    }
+
+    if (f1.order() != f2.order()) {
+      return false;
+    }
+
+    // Check if both have default values
+    if (f1.hasDefaultValue() != f2.hasDefaultValue()) {
+      return false;
+    }
+
+    // If both have default values, they must be equal
+    if (f1.hasDefaultValue() && !f1.defaultVal().get().equals(f2.defaultVal().get())) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean fieldEquals(HoodieSchemaField f1, HoodieSchemaField f2) {
+    if (!validateField(f1, f2)) {
+      return false;
+    }
+
+    return schemaEqualsInternal(f1.schema(), f2.schema());
+  }
+
+  protected boolean enumSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    // Check name equality first
+    if (!s1.getName().equals(s2.getName())) {
+      return false;
+    }
+
+    List<String> symbols1 = s1.getEnumSymbols();
+    List<String> symbols2 = s2.getEnumSymbols();
+
+    // Quick size check before creating sets
+    if (symbols1.size() != symbols2.size()) {
+      return false;
+    }
+
+    return symbols1.equals(symbols2);
+  }
+
+  protected boolean unionSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    List<HoodieSchema> types1 = s1.getTypes();
+    List<HoodieSchema> types2 = s2.getTypes();
+
+    if (types1.size() != types2.size()) {
+      return false;
+    }
+
+    // Create sets of effectively equal types
+    Set<SchemaWrapper> set1 = types1.stream().map(SchemaWrapper::new).collect(Collectors.toSet());
+    Set<SchemaWrapper> set2 = types2.stream().map(SchemaWrapper::new).collect(Collectors.toSet());
+
+    // Compare sets instead of ordered lists
+    return set1.equals(set2);
+  }
+
+  private boolean arraySchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    return schemaEqualsInternal(s1.getElementType(), s2.getElementType());
+  }
+
+  private boolean mapSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    return schemaEqualsInternal(s1.getValueType(), s2.getValueType());
+  }
+
+  protected boolean validateFixed(HoodieSchema s1, HoodieSchema s2) {
+    return s1.getName().equals(s2.getName()) && s1.getFixedSize() == s2.getFixedSize();
+  }
+
+  private boolean fixedSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    if (!validateFixed(s1, s2)) {
+      return false;
+    }
+    return logicalTypeSchemaEquals(s1, s2);
+  }
+
+  private static boolean primitiveSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    // For primitive types, just check logical type
+    return logicalTypeSchemaEquals(s1, s2);
+  }
+
+  private static boolean logicalTypeSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
```

Comment:
I think we should break these up into functions for each type so we can directly call them from `schemaEqualsInternal`